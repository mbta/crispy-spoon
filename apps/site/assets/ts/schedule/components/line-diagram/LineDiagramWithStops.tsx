import React, { ReactElement } from "react";
import { hasBranchLines } from "./line-diagram-helpers";
import Diagram from "./graphics/Diagram";
import StopListWithBranches from "./StopListWithBranches";
import { CommonLineDiagramProps } from "./__line-diagram";
import useStopPositions, { RefList } from "./graphics/useStopPositions";
import StopCard from "./StopCard";

export const StopRefContext = React.createContext<[RefList, () => void]>([
  {},
  () => {}
]);

const LineDiagramWithStops = (
  props: CommonLineDiagramProps
): ReactElement<HTMLElement> => {
  const { stops, handleStopClick, liveData } = props;

  // create a ref for each stop - we will use this to track the location of the stop so we can place the line diagram bubbles
  const [stopRefsMap, updateAllStopCoords] = useStopPositions(stops);

  return (
    <StopRefContext.Provider value={[stopRefsMap, updateAllStopCoords]}>
      <ol className="m-schedule-diagram">
        <Diagram lineDiagram={stops} liveData={liveData} />
        {hasBranchLines(stops) ? (
          <StopListWithBranches
            stops={stops}
            handleStopClick={handleStopClick}
            liveData={liveData}
          />
        ) : (
          stops.map(stop => (
            <StopCard
              key={stop.route_stop.id}
              stop={stop}
              onClick={handleStopClick}
              liveData={liveData[stop.route_stop.id]}
            />
          ))
        )}
      </ol>
    </StopRefContext.Provider>
  );
};

export default LineDiagramWithStops;
