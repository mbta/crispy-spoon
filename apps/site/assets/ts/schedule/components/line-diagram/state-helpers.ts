import { AnyAction, Reducer, Store, createStore } from "redux";
import { LineDiagramState, LineDiagramStateAction } from "./__line-diagram";
import { LineDiagramStop } from "../__schedule";

export type StopCoord = [number, number];
export interface CoordState {
  [key: string]: StopCoord | null;
}

export interface CoordAction extends AnyAction {
  stop: string;
  coords: StopCoord | null;
}

export const coordReducer: Reducer<CoordState, CoordAction> = (
  allCoordinates: CoordState | undefined,
  action: CoordAction
): CoordState => {
  const { stop, coords } = action;
  if (stop) {
    return {
      ...allCoordinates,
      [stop]: coords
    };
  }

  return allCoordinates || {};
};

export const createLineDiagramCoordStore = (
  lineDiagram: LineDiagramStop[]
): Store<CoordState, CoordAction> => {
  const initialCoordState = lineDiagram
    .map(ld => ld.route_stop.id)
    .reduce((o, s) => ({ ...o, [s]: null }), {});
  const store = createStore(
    coordReducer,
    initialCoordState,
    /* eslint-disable no-underscore-dangle, @typescript-eslint/no-explicit-any */
    (window as any).__REDUX_DEVTOOLS_EXTENSION__ &&
      (window as any).__REDUX_DEVTOOLS_EXTENSION__()
    /* eslint-enable no-underscore-dangle, @typescript-eslint/no-explicit-any */
  );

  return store;
};

export const lineDiagramReducer = (
  state: LineDiagramState,
  action: LineDiagramStateAction
): LineDiagramState => {
  switch (action.type) {
    case "initialize":
      return {
        ...state,
        direction: action.direction,
        origin: action.origin,
        modalMode: "schedule",
        modalOpen: true
      };
    case "set_direction":
      return {
        ...state,
        direction: action.direction
      };
    case "set_origin":
      if (action.origin) {
        // back to schedule modal
        return {
          ...state,
          origin: action.origin,
          modalMode: "schedule"
        };
      }
      // origin modal (but don't overwrite current origin)
      return {
        ...state,
        modalMode: "origin"
      };
    case "toggle_modal":
      return {
        ...state,
        modalOpen: action.modalOpen
      };
    default:
      return state;
  }
};
