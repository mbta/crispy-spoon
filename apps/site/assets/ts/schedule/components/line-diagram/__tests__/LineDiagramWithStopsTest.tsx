import React from "react";
import * as redux from "react-redux";
import { mount, ReactWrapper } from "enzyme";
import { cloneDeep, merge } from "lodash";
import { RouteType } from "../../../../__v3api";
import { LineDiagramStop } from "../../__schedule";
import simpleLineDiagram from "./lineDiagramData/simple.json"; // not a full line diagram
import outwardLineDiagram from "./lineDiagramData/outward.json"; // not a full line diagram
import simpleLiveData from "./lineDiagramData/live-data.json";
import LineDiagramWithStops from "../LineDiagramWithStops";
import { createLineDiagramCoordStore } from "../state-helpers";
import StopListWithBranches from "../StopListWithBranches";

const lineDiagram = (simpleLineDiagram as unknown) as LineDiagramStop[];
let lineDiagramBranchingOut = (outwardLineDiagram as unknown) as LineDiagramStop[];

const route = {
  type: 3 as RouteType,
  name: "route 1",
  long_name: "route 1 long name",
  color: "F00B42",
  id: "route-1",
  direction_names: {
    0: "Outbound",
    1: "Inbound"
  },
  direction_destinations: {
    0: "Begin",
    1: "End"
  },
  description: "key_bus_route",
  "custom_route?": false,
  header: "",
  alerts: []
};

lineDiagram.forEach(({ route_stop }) => {
  route_stop.route = cloneDeep(route);
});

lineDiagramBranchingOut.forEach(({ route_stop }) => {
  route_stop.route = cloneDeep(route);
});

let lineDiagramBranchingIn = cloneDeep(lineDiagramBranchingOut).reverse();
const CRroute = merge(cloneDeep(route), { type: 2 as RouteType });
lineDiagramBranchingIn.forEach(({ route_stop }) => {
  route_stop.route = CRroute;
  if (route_stop["is_terminus?"]) {
    route_stop["is_beginning?"] = !route_stop["is_beginning?"];
  }
});

const handleStopClick = () => {};
const liveData = {};
const store = createLineDiagramCoordStore(lineDiagram);

describe("LineDiagramWithStops", () => {
  let wrapper: ReactWrapper;
  beforeEach(() => {
    wrapper = mount(
      <redux.Provider store={store}>
        <LineDiagramWithStops
          stops={lineDiagram}
          handleStopClick={handleStopClick}
          liveData={liveData}
        />
      </redux.Provider>
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it("renders and matches snapshot", () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("shows <StopListWithBranches /> if the line has branches", () => {
    // wrapper's line diagram has no branches
    expect(wrapper.find(StopListWithBranches)).toHaveLength(0);
    const wrapperWithBranches = mount(
      <redux.Provider store={store}>
        <LineDiagramWithStops
          stops={lineDiagramBranchingOut}
          handleStopClick={handleStopClick}
          liveData={liveData}
        />
      </redux.Provider>
    );
    expect(wrapperWithBranches.find(StopListWithBranches)).toHaveLength(1);
  });
});
