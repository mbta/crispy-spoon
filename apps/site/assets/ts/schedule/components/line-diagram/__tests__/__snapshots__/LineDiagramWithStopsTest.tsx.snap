// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineDiagramWithStops renders and matches snapshot 1`] = `
"<Provider store={{...}}>
  <LineDiagramWithStops stops={{...}} handleStopClick={[Function: handleStopClick]} liveData={{...}}>
    <ol className=\\"m-schedule-diagram\\">
      <Diagram lineDiagram={{...}} liveData={{...}}>
        <svg xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" aria-labelledby=\\"diagram-title diagram-desc\\" className=\\"line-diagram-svg bus\\" width=\\"28px\\" height=\\"100%\\">
          <title id=\\"diagram-title\\">
            Line diagram for
             
            route 1
          </title>
          <desc id=\\"diagram-desc\\">
            4 stops going Outbound to Begin
          </desc>
          <Line from={{...}} to={{...}}>
            <line className=\\"line-diagram-svg__line\\" strokeWidth=\\"8px\\" x1=\\"9px\\" x2=\\"9px\\" y1=\\"30px\\" y2=\\"50px\\" />
          </Line>
          <Line from={{...}} to={{...}}>
            <line className=\\"line-diagram-svg__line\\" strokeWidth=\\"8px\\" x1=\\"9px\\" x2=\\"9px\\" y1=\\"50px\\" y2=\\"70px\\" />
          </Line>
          <Line from={{...}} to={{...}}>
            <line className=\\"line-diagram-svg__line\\" strokeWidth=\\"8px\\" x1=\\"9px\\" x2=\\"9px\\" y1=\\"70px\\" y2=\\"90px\\" />
          </Line>
          <Stop stop={{...}}>
            <circle className=\\"line-diagram-svg__stop\\" r=\\"4px\\" cx=\\"10px\\" cy=\\"30px\\" />
          </Stop>
          <Stop stop={{...}}>
            <circle className=\\"line-diagram-svg__stop\\" r=\\"4px\\" cx=\\"10px\\" cy=\\"50px\\" />
          </Stop>
          <Stop stop={{...}}>
            <circle className=\\"line-diagram-svg__stop\\" r=\\"4px\\" cx=\\"10px\\" cy=\\"70px\\" />
          </Stop>
          <Stop stop={{...}}>
            <circle className=\\"line-diagram-svg__stop\\" r=\\"4px\\" cx=\\"10px\\" cy=\\"90px\\" />
          </Stop>
        </svg>
      </Diagram>
      <StopCard stop={{...}} onClick={[Function: handleStopClick]} liveData={[undefined]}>
        <li className=\\"m-schedule-diagram__stop\\" style={{...}}>
          <section className=\\"m-schedule-diagram__content\\">
            <header className=\\"m-schedule-diagram__stop-heading\\">
              <h4 className=\\"m-schedule-diagram__stop-link\\">
                <a href=\\"/stops/line-origin\\">
                  <MatchHighlight text=\\"Origin\\" matchQuery={[undefined]}>
                    <span>
                      Origin
                    </span>
                  </MatchHighlight>
                </a>
              </h4>
              <div className=\\"m-schedule-diagram__features\\">
                <Component tooltipText=\\"Parking\\" tooltipOptions={{...}}>
                  <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={true} data-html={false} data-selector=\\"true\\" data-original-title=\\"Parking\\" title=\\"Parking\\">
                    <span className=\\"notranslate c-svg__icon-parking-default m-schedule-diagram__feature-icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                  </span>
                </Component>
              </div>
            </header>
            <div className=\\"m-schedule-diagram__stop-details\\">
              <div className=\\"m-schedule-diagram__connections\\" />
            </div>
            <footer className=\\"m-schedule-diagram__footer\\">
              <button className=\\"btn btn-link\\" type=\\"button\\" onClick={[Function: onClick]}>
                View schedule
              </button>
            </footer>
          </section>
        </li>
      </StopCard>
      <StopCard stop={{...}} onClick={[Function: handleStopClick]} liveData={[undefined]}>
        <li className=\\"m-schedule-diagram__stop\\" style={{...}}>
          <section className=\\"m-schedule-diagram__content\\">
            <header className=\\"m-schedule-diagram__stop-heading\\">
              <h4 className=\\"m-schedule-diagram__stop-link\\">
                <a href=\\"/stops/line-stop1\\">
                  <span className=\\"notranslate c-svg__icon-alerts-triangle\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                  <span className=\\"sr-only\\">
                    Service alert or delay
                  </span>
                   
                  <MatchHighlight text=\\"Stop 1\\" matchQuery={[undefined]}>
                    <span>
                      Stop 1
                    </span>
                  </MatchHighlight>
                </a>
              </h4>
              <div className=\\"m-schedule-diagram__features\\" />
            </header>
            <div className=\\"m-schedule-diagram__stop-details\\">
              <div className=\\"m-schedule-diagram__connections\\">
                <Component href=\\"/schedules/Orange/line\\" tooltipText=\\"Orange Line\\" tooltipOptions={{...}}>
                  <a href=\\"/schedules/Orange/line\\" data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={false} data-html={false} data-selector=\\"true\\" data-original-title=\\"Orange Line\\" title=\\"Orange Line\\">
                    <span className=\\"m-schedule-diagram__connection\\">
                      <span className=\\"notranslate c-svg__icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                    </span>
                  </a>
                </Component>
                <Component href=\\"/schedules/Green-C/line\\" tooltipText=\\"Green Line C\\" tooltipOptions={{...}}>
                  <a href=\\"/schedules/Green-C/line\\" data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={false} data-html={false} data-selector=\\"true\\" data-original-title=\\"Green Line C\\" title=\\"Green Line C\\">
                    <span className=\\"m-schedule-diagram__connection\\">
                      <span className=\\"notranslate c-svg__icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                    </span>
                  </a>
                </Component>
              </div>
            </div>
            <footer className=\\"m-schedule-diagram__footer\\">
              <button className=\\"btn btn-link\\" type=\\"button\\" onClick={[Function: onClick]}>
                View schedule
              </button>
            </footer>
          </section>
        </li>
      </StopCard>
      <StopCard stop={{...}} onClick={[Function: handleStopClick]} liveData={[undefined]}>
        <li className=\\"m-schedule-diagram__stop\\" style={{...}}>
          <section className=\\"m-schedule-diagram__content\\">
            <header className=\\"m-schedule-diagram__stop-heading\\">
              <h4 className=\\"m-schedule-diagram__stop-link\\">
                <a href=\\"/stops/line-stop2\\">
                  <span className=\\"notranslate c-svg__icon-alerts-triangle\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                  <span className=\\"sr-only\\">
                    Service alert or delay
                  </span>
                   
                  <MatchHighlight text=\\"Stop 2\\" matchQuery={[undefined]}>
                    <span>
                      Stop 2
                    </span>
                  </MatchHighlight>
                </a>
              </h4>
              <div className=\\"m-schedule-diagram__features\\">
                <Component tooltipText=\\"Accessible\\" tooltipOptions={{...}}>
                  <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={true} data-html={false} data-selector=\\"true\\" data-original-title=\\"Accessible\\" title=\\"Accessible\\">
                    <span className=\\"notranslate c-svg__icon-acessible-default m-schedule-diagram__feature-icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                  </span>
                </Component>
              </div>
            </header>
            <div className=\\"m-schedule-diagram__stop-details\\">
              <div className=\\"m-schedule-diagram__connections\\">
                <Component href=\\"/schedules/62/line\\" tooltipText=\\"Route 62\\" tooltipOptions={{...}}>
                  <a href=\\"/schedules/62/line\\" data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={false} data-html={false} data-selector=\\"true\\" data-original-title=\\"Route 62\\" title=\\"Route 62\\">
                    <span className=\\"c-icon__bus-pill--small m-schedule-diagram__connection u-bg--bus\\">
                      62
                    </span>
                  </a>
                </Component>
                <Component href=\\"/schedules/67/line\\" tooltipText=\\"Route 67\\" tooltipOptions={{...}}>
                  <a href=\\"/schedules/67/line\\" data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={false} data-html={false} data-selector=\\"true\\" data-original-title=\\"Route 67\\" title=\\"Route 67\\">
                    <span className=\\"c-icon__bus-pill--small m-schedule-diagram__connection u-bg--bus\\">
                      67
                    </span>
                  </a>
                </Component>
              </div>
            </div>
            <footer className=\\"m-schedule-diagram__footer\\">
              <button className=\\"btn btn-link\\" type=\\"button\\" onClick={[Function: onClick]}>
                View schedule
              </button>
            </footer>
          </section>
        </li>
      </StopCard>
      <StopCard stop={{...}} onClick={[Function: handleStopClick]} liveData={[undefined]}>
        <li className=\\"m-schedule-diagram__stop\\" style={{...}}>
          <section className=\\"m-schedule-diagram__content\\">
            <header className=\\"m-schedule-diagram__stop-heading\\">
              <h4 className=\\"m-schedule-diagram__stop-link\\">
                <a href=\\"/stops/line-destination\\">
                  <MatchHighlight text=\\"Destination\\" matchQuery={[undefined]}>
                    <span>
                      Destination
                    </span>
                  </MatchHighlight>
                </a>
              </h4>
              <div className=\\"m-schedule-diagram__features\\">
                <Component tooltipText=\\"Parking\\" tooltipOptions={{...}}>
                  <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={true} data-html={false} data-selector=\\"true\\" data-original-title=\\"Parking\\" title=\\"Parking\\">
                    <span className=\\"notranslate c-svg__icon-parking-default m-schedule-diagram__feature-icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                  </span>
                </Component>
                <Component tooltipText=\\"Accessible\\" tooltipOptions={{...}}>
                  <span href={[undefined]} data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={true} data-html={false} data-selector=\\"true\\" data-original-title=\\"Accessible\\" title=\\"Accessible\\">
                    <span className=\\"notranslate c-svg__icon-acessible-default m-schedule-diagram__feature-icon\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                  </span>
                </Component>
              </div>
            </header>
            <div className=\\"m-schedule-diagram__stop-details\\">
              <div className=\\"m-schedule-diagram__connections\\">
                <Component href=\\"/schedules/BOAT-A/line\\" tooltipText=\\"Atlantis\\" tooltipOptions={{...}}>
                  <a href=\\"/schedules/BOAT-A/line\\" data-toggle=\\"tooltip\\" data-trigger=\\"hover focus\\" data-placement=\\"bottom\\" data-animation={false} data-html={false} data-selector=\\"true\\" data-original-title=\\"Atlantis\\" title=\\"Atlantis\\">
                    <span className=\\"m-schedule-diagram__connection\\">
                      <span className=\\"notranslate c-svg__icon-bus-small\\" aria-hidden=\\"false\\" dangerouslySetInnerHTML={{...}} />
                    </span>
                  </a>
                </Component>
              </div>
            </div>
            <footer className=\\"m-schedule-diagram__footer\\">
              <button className=\\"btn btn-link\\" type=\\"button\\" onClick={[Function: onClick]}>
                View schedule
              </button>
            </footer>
          </section>
        </li>
      </StopCard>
    </ol>
  </LineDiagramWithStops>
</Provider>"
`;
